File: data/readable_format/fortran/file892.f90
program digit_factorials
    implicit none
    integer :: max_digit

    ! Function and subroutine declarations are moved to a contains section below
    
    ! Call tests and solver here, outside any contains section
    call test_factorial()
    call test_digit_sequence()
    call solver

contains

    function digit_sequence(n) result(d)
        integer, intent(in) :: n
        integer, allocatable :: d(:)
        integer :: a, i

        allocate(d(10)) ! Assuming a maximum of 10 digits for simplicity
        d = 0
        a = n
        i = 0
        do while (a > 0)
            i = i + 1
            d(i) = mod(a, 10)
            a = a / 10
        end do
        d = d(1:i)
    end function digit_sequence

    function factorial(n) result(f)
        integer, intent(in) :: n
        integer :: f
        integer :: i

        f = 1
        if (n > 0) then
            do i = 2, n
                f = f * i
            end do
        end if
    end function factorial

    subroutine solver
        integer :: n, p, q, i, j, a, sum
        integer, allocatable :: seq(:)

        p = factorial(9)
        n = 1
        do while (p * n > 10 ** (n - 1))
            n = n + 1
        end do
        max_digit = n

        sum = 0
        do i = 3, p * max_digit
            seq = digit_sequence(i)
            a = 0
            do j = 1, size(seq)
                q = factorial(seq(j))
                a = a + q
            end do
            if (a == i) then
                sum  = sum + a
                print *, i
            end if
        end do

        print *, 'result =', sum

    end subroutine solver

    subroutine test_factorial()
        integer :: result
        result = factorial(5)
        if (result /= 120) then
            print *, 'Factorial test failed'
        else
            print *, 'Factorial test passed'
        end if
    end subroutine test_factorial

    subroutine test_digit_sequence()
        integer, allocatable :: seq(:)
        integer :: n
        n = 145
        seq = digit_sequence(n)
        if (size(seq) == 3 .and. seq(1) == 5 .and. seq(2) == 4 .and. seq(3) == 1) then
            print *, 'Digit sequence test passed'
        else
            print *, 'Digit sequence test failed'
        end if
    end subroutine test_digit_sequence

end program digit_factorials

File: data/readable_format/cpp/file892.cpp
#include <iostream>
#include <vector>
#include <cassert> // For assert()
#include <cmath>    // For std::pow()

// Function to calculate factorial
int factorial(int n) {
    int result = 1;
    for (int i = 1; i <= n; ++i) {
        result *= i;
    }
    return result;
}

// Function to generate the sequence of digits of a number
std::vector<int> digit_sequence(int n) {
    std::vector<int> digits;
    while (n > 0) {
        digits.push_back(n % 10);
        n /= 10;
    }
    return digits;
}

// Solver function implementing the logic to find and sum the special numbers
void solver() {
    int max_digit = 7; // Arbitrary choice for max digits
    int p = factorial(9);
    int n = 1;
    while (p * n > std::pow(10, n - 1)) {
        n = n + 1;
    }
    max_digit = n;

    int sum = 0;
    for (int i = 10; i <= p * max_digit; ++i) { // Starting from 10, as single-digit numbers are not considered
        auto digits = digit_sequence(i);
        int digit_sum = 0;
        for (auto digit : digits) {
            digit_sum += factorial(digit);
        }
        if (digit_sum == i) {
            sum += i;
            std::cout << i << std::endl;
        }
    }
    std::cout << "Result = " << sum << std::endl;
}

// Test function for factorial
void test_factorial() {
    assert(factorial(5) == 120);
    std::cout << "Factorial test passed." << std::endl;
}

// Test function for digit sequence
void test_digit_sequence() {
    auto seq = digit_sequence(145); // Modified to remove unnecessary parameter
    assert(seq.size() == 3 && seq[0] == 5 && seq[1] == 4 && seq[2] == 1);
    std::cout << "Digit sequence test passed." << std::endl;
}

int main() {
    test_factorial();
    test_digit_sequence();
    solver();
    return 0;
}